experiment/compatibility-versions/validate-compatibility-versions-feature-gates.sh

```diff
@@ -0,0 +1,245 @@
+    if [[ "$stage" == "ALPHA" ]]; then
+      continue
+    fi
+
+    if [[ -z "$got" ]]; then
+      # Missing from metrics
+      if [[ "$locked" == "true" ]]; then
+        continue
+      fi
+      echo "FAIL: expected feature gate '$feature_name' not found in metrics (stage=${stage}, lockToDefault=${locked})" \
+      >> "${results_file}"
+      continue
+    fi
+
+    # If present, stage!=ALPHA => compare true/false enabled value
+    if [[ "$got" != "$want" ]]; then
+      echo "FAIL: feature '$feature_name' expected value $want, got $got" \
+        >> "${results_file}"
+    fi
+  done
+
+  # For each "expected_unversioned" feature:
+  # - If missing from /metrics => fail unless stage==ALPHA or lock==true
+  # - If present => compare numeric value
+  for unversioned_feature_name in "${!expected_unversioned_stage[@]}"; do
+    local unversioned_stage="${expected_unversioned_stage[$unversioned_feature_name]}"
+    local unversioned_locked="${expected_unversioned_lock[$unversioned_feature_name]}"
+    local unversioned_want="${expected_unversioned_value[$unversioned_feature_name]}"
+
+    local got="${actual_features[$unversioned_feature_name]:-}"  # empty if missing
+
+    # If present, but stage==ALPHA => no checks are done
+    if [[ "$unversioned_stage" == "ALPHA" ]]; then
+      continue
+    fi
+
+    if [[ -z "$got" ]]; then
+      # Missing from metrics
+      if [[ "$unversioned_locked" == "true" ]]; then
+        continue
+      fi
+      echo "FAIL: expected unversioned feature gate '$unversioned_feature_name' not found in metrics (lockToDefault=${unversioned_locked})" \
+      >> "${results_file}"
+      continue
+    fi
+
+    # If present, compare true/false enabled value
+    if [[ "$got" != "$unversioned_want" ]]; then
+      echo "FAIL: unversioned feature '$unversioned_feature_name' expected value $unversioned_want, got $got" \
+        >> "${results_file}"
+    fi
+  done
+
+  # For each actual feature in /metrics not in the "expected" maps (versioned OR unversioned),
+  #  - if it's "1", we fail as "unexpected feature". because new gates not found in previous
+  #    expected gates can only be introduced if they are off by default (0) but not on by default (1)
+    #    NOTE: if the new feature is a client-go feature then we do not fail but continue
+  for feature_name in "${!actual_features[@]}"; do
+    if [[ -z "${expected_stage[$feature_name]:-}" ]] && [[ -z "${expected_unversioned_stage[$feature_name]:-}" ]]; then
+      local got="${actual_features[$feature_name]}"
+      if [[ "$got" == "1" ]]; then
+        # Check to see if gate is found in client-go and if so, continue
+        if grep -q "$feature_name" staging/src/k8s.io/client-go/features/known_features.go; then
+          continue
+        fi
+        echo "FAIL: unexpected feature '$feature_name' found in /metrics, got=1" \
+          >> "${results_file}"
+      fi
+    fi
+  done
+}
+
+# --- Main execution when script is called directly ---
```

Comment:
I would recommend just only supporting calling and condense this down to a single codepath / exec it from the other script instead of trying to support both exec and sourcing.

Reply:
(up to you, but I'd suggest considering this before it gets wider usage, feel free to /hold cancel and punt)

Reply:
Ah makes sense, I've updated the script to support only being exec'd for now.